feat(all): Unify inode numbers type

lgsilva3087 opened the pull request:
SaunaFS uses uint32_t as type for inode numbers and quantities. It means, a cluster can not grow beyond about 4 billion files.

The inode numbers type (uint32_t) has different definitions across the code, like:
- typedef uint32_t SaunaClient::Inode
- using InodeId = uint32_t; // In the ACLs
- plain uint32_t
- etc.

Also, in several places, where the inode number is serialized, the fixed size of 4 bytes is added to the buffer size, causing problems if we change the type to be uint64_t for instance (8 bytes).

This change starts to unify the type for inode numbers across the complete codebase. The new type (inode_t) is now defined in the new file type_defs.h in a way compatible with C++ and C (for the client lib). All references to an inode number should use this type.

The main changes introduced in this PR are:

- inode_t as type for inode numbers and quantities.
- ENABLE_INODE64 option in the main CMakeFile.txt (experimental, defaults to 32 bits).
- New PRIiNode macro for legacy logging.
- New type safe putINode and getINode functions in datapack.h.
- Templated type safe put32bit and get32bit to avoid incorrect usage.
- Massive reduction of magic numbers used in the protocol and in metadata, to allow easy switching between 32 and 64 bits inodes.

This change does not add support for 64 bits inodes in SaunaFS, but prepares the code to start adding it.

lgsilva3087 committed:
feat(all): Unify inode numbers type

The inode numbers type (currently uint32_t) is not easily replaceable,
there are different kinds of definitions across the code, like:
- typedef uint32_t SaunaClient::Inode
- using InodeId = uint32_t; // In the ACLs
- plain uint32_t
- etc.

This commit starts to unify the type for inode numbers across the
complete codebase. The type is now defined in the new file type_defs.h
in a way compatible with C++ and C (for the client lib). All other
references to an inode number should use this type (Inode).

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 commented on src/mount/client/sauna_client_c_linkage.h:
```diff
@@ -38,40 +38,40 @@ extern "C" {
 
 int saunafs_readdir(SaunaClient::Context &ctx, uint64_t opendirSessionID,
-                    SaunaClient::Inode ino, off_t off, size_t max_entries,
+                    Inode ino, off_t off, size_t max_entries,
```
Yes, that was my change.

lgsilva3087 commented on src/mount/client/sauna_client_c_linkage.h:
```diff
@@ -38,40 +38,40 @@ extern "C" {
 
 int saunafs_read_special_inode(SaunaClient::Context &ctx,
-                               SaunaClient::Inode ino, size_t size, off_t off,
+                               Inode ino, size_t size, off_t off,
```
Yes, that's part of the change.

lgsilva3087 commented on src/admin/manage_locks_command.cc:
```diff
@@ -111,12 +112,12 @@ static void processUnlock(RegisteredAdminConnection &conn, const Options &option
 	for (uint64_t begin = 0; true; begin += kDataPortion) {
 		MessageBuffer msg;
-		if (inode) {
+		if (inode != 0U) {
```
Nothing fancy, to avoid the clang-tidy warning about implicit check.

lgsilva3087 commented on src/mount/client/sauna_client_c_linkage.h:
```diff
@@ -83,45 +83,45 @@ int saunafs_readtrash(SaunaClient::Context &ctx,
+int saunafs_rename(SaunaClient::Context &ctx, Inode parent, const char *name,
+	            Inode newparent, const char *newname);
 int saunafs_statfs(uint64_t *totalspace, uint64_t *availspace, uint64_t *trashspace,
 	             uint64_t *reservedspace, uint32_t *inodes);
```
Thanks, done.

antuan96314 commented on src/common/type_defs.h:
```diff
@@ -0,0 +1,34 @@
+#ifdef __cplusplus
+#include <cstdint>
+
+using Inode = uint32_t;
```
Consider naming the type as `Inode_t`
```suggestion
using Inode_t = uint32_t;
```

lgsilva3087 replied:
Thanks, yes, I will do it in later commits.

lgsilva3087 committed:
feat(ganesha): Improve wire_to_host for 64 bits inodes

The function was extended to handle 32 or 64 bits inodes.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(all): Add the PRIiNode macro

The PRIiInode macro depends on the define inode_t type and should be
used as a replacement for '% PRIu32' or '%u' or similars.

PRIXiNode was also added for hex printing.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(all): Add the putINode function

putINode should be used to serialize inode numbers into buffers. It
uses the correct size for 32 or 64 bits.

As a side effect, the function put32bit was improved to detect
incorrect usage at compile time.

Some places with magic numbers, like 4 for inode number sizes, were
replaced by kinode_t_size or sizeof(inode_t), which points to the
correct size.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(all): Increase type safety of get32bit

The new version checks for the correct type size at compile time. This
behavior helps by early detecting possible incorrect usages.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(all): Add the getINode type aware function

Inode related variables should be deserialized using the new getINode
function, which depends on the selected inode size (32 or 64 bits).

As a side effect, numerous magic numbers for buffer sizes were replaced
by better named constants or by compile time calculated values using
the correct `sizeof` of the involved parameters.

Co-authored-by: Crash <crash@leil.io>

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(all): Reduce magic numbers

The magic numbers are causing protocol and metadata errors when inode_t
points to 64 bits. Some of the errors are related to:

- Buffer's sizes.
- Protocol packets expected and sent sizes.
- Metadata file offsets.

Some unused functions were found and commented for later removal.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
fix(mds): Fix metadata save/load to file

Fix errors related to metadata serialization and deserialization,
mostly for 64 bits inodes.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(all): Add the option ENABLE_INODE64 to cmake

The new option controls the size of inode numbers type (inode_t).

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
fix(mds): Fix build after rebase

Signed-off-by: guillex <guillex@leil.io>

ralcolea commented on src/master/restore.cc:
```diff
@@ -252,7 +255,8 @@ int do_checksum(const char *filename, uint64_t lv, uint32_t, const char *ptr) {
 int do_create(const char *filename, uint64_t lv, uint32_t ts, const char *ptr) {
-	uint32_t parent,mode,uid,gid,rdev,inode;
+	inode_t parent;
+	uint32_t mode,uid,gid,rdev,inode;
```
`inode` type should change to `inode_t`, like `parent`?

lgsilva3087 replied:
Thanks. Done.

dmga44 commented on src/master/chunks.cc:
```diff
@@ -887,7 +888,8 @@ void chunk_store_chunkcounters(uint8_t *buff,uint8_t matrixid) {
 		for (int i = 0; i < CHUNK_MATRIX_SIZE; i++) {
 			for (int j = 0; j < CHUNK_MATRIX_SIZE; j++) {
-				put32bit(&buff, Chunk::allFullChunkCopies[i][j]);
+				// TODO(Guillex): possible truncation if the chunks number is greater than 2^32
+				put32bit(&buff, static_cast<uint32_t>(Chunk::allFullChunkCopies[i][j]));
```
As you point out, this needs to be remembered for an organic support to 64 bits inodes.

dmga44 commented on src/master/datacachemgr.cc:
```diff
@@ -36,18 +38,18 @@
+static inode_t dcm_lru_first,dcm_lru_last;
 
-int dcm_open(uint32_t inode,uint32_t sessionid) {
+int dcm_open(inode_t inode, uint32_t sessionid) {
 	uint32_t ih = DCM_INODE_HASH(inode);
 	uint32_t p,pp,np;
 	p = dcm_inodehash[ih];
```
About these lines:
- please check that **DCM_INODE_HASH** multiplication could overflow (`uint128_t` ?)
- `p`, `pp` and `np` variables should be also `inode_t`.

lgsilva3087 replied:
Solved.

dmga44 commented on src/master/datacachemgr.cc:
```diff
@@ -132,7 +134,7 @@ int dcm_open(uint32_t inode,uint32_t sessionid) {
-void dcm_access(uint32_t inode,uint32_t sessionid) {
+void dcm_access(inode_t inode, uint32_t sessionid) {
 	uint32_t ih = DCM_INODE_HASH(inode);
 	uint32_t p,pp,np;
 	p = dcm_inodehash[ih];
```
Same here (please check rest of the file).

lgsilva3087 replied:
Solved.

dmga44 commented on src/common/special_inode_defs.h:
```diff
@@ -22,9 +22,16 @@
-#define SPECIAL_INODE_BASE   0xFFFFFFF0U
+#include "common/type_defs.h"
+
+#ifdef SAUNAFS_USE_INODE64
+#define SPECIAL_INODE_BASE 0xFFFFFFF0LU
+#define SPECIAL_INODE_ROOT 0x01LU
```
For 64 bit inodes, it would be necessary to increase the value of `SPECIAL_INODE_BASE` because this value would collide with regular supported inodes (note it is less than 2^32).

lgsilva3087 replied:
I added the TODO for 64 bits as a reminder.

dmga44 commented on src/master/filesystem_metadata.h:
```diff
@@ -42,7 +42,7 @@ struct FilesystemMetadata {
 	xattr_inode_entry *xattr_inode_hash[XATTR_INODE_HASH_SIZE];
 	xattr_data_entry *xattr_data_hash[XATTR_DATA_HASH_SIZE];
-	IdPoolDetainer<uint32_t, uint32_t> inode_pool;
+	IdPoolDetainer<inode_t, uint32_t> inode_pool;
```
The implications of increasing the `SPECIAL_INODE_BASE` constant and `MAX_REGULAR_INODE`, could modify the behavior of this `inode_pool`. We'd need to check the implementation of this structure.

lgsilva3087 replied:
Agree, but so far for 32 bits (which is the only one supported so far) is not affected.
Adding the TODO for 64 bits support.

dmga44 reviewed (commented):
Review part 1/TBD

dmga44 commented on src/master/filesystem_node.h:
```diff
@@ -28,17 +28,9 @@
 namespace detail {
 
-inline FSNode *fsnodes_id_to_node_internal(uint32_t id) {
+inline FSNode *fsnodes_id_to_node_internal(inode_t id) {
 	FSNode *p;
 	uint32_t nodepos = NODEHASHPOS(id);
 	for (p = gMetadata->nodehash[nodepos]; p; p = p->next) {
```
When the number of inodes grows even larger (4 billion +), this loop would be very expensive and it is being called everywhere. Suggest to replace the linked list for a more efficient structure for the final version of the system supporting 64 bits inodes.

lgsilva3087 replied:
Agree.

dmga44 commented on src/master/filesystem_xattr.h:
```diff
@@ -69,7 +69,7 @@ static inline int xattr_namecheck(uint8_t anleng, const uint8_t *attrname) {
 #endif /* METARESTORE */
 
-static inline uint32_t xattr_data_hash_fn(uint32_t inode, uint8_t anleng, const uint8_t *attrname) {
+static inline uint32_t xattr_data_hash_fn(inode_t inode, uint8_t anleng, const uint8_t *attrname) {
 	uint32_t hash = inode * 5381U;
 	while (anleng) {
 		hash = (hash * 33U) + (*attrname);
```
Though the overflows in this function are on purpose, what I noticed here is that the `xattr_data_hash` array of size `XATTR_DATA_HASH_SIZE` is very small. It means that for a storage with a huge number of xattr the operations like `xattr_setattr` and `xattr_getattr` whould be very expensive.

lgsilva3087 replied:
Ack

dmga44 commented on src/master/masterconn.cc:
```diff
@@ -636,7 +636,8 @@ void MasterConn::downloadNext() {
 			put32bit(&ptr, kMetadataDownloadBlocksize);
 		} else {
-			put32bit(&ptr, fileSize - downloadOffset);
+			// TODO(Guillex): Possible truncation if the file size is too big
+			put32bit(&ptr, static_cast<uint32_t>(fileSize - downloadOffset));
```
Truncate won't happen because `fileSize - downloadOffset` is less than or equal to `kMetadataDownloadBlocksize`. Truncation could happen if `kMetadataDownloadBlocksize` is bigger than 2^32.

lgsilva3087 replied:
Agree, TODO removed.

ralcolea commented on src/master/chunks.cc:
```diff
@@ -1806,6 +1808,19 @@ void chunk_got_duptrunc_status(matocsserventry *ptr, uint64_t chunkId, ChunkPart
+uint32_t get_chunk_info_serialized_size() {
+	constexpr uint32_t kInfoSize =
+	    sizeof(chunksinfo_loopstart) + sizeof(chunksinfo_loopend) +
+	    sizeof(job_info::del_invalid) + sizeof(job_info::del_invalid) +
```
Maybe the terms sumed twice can be replaced by 2 times the duplicated term:
```cpp
constexpr uint32_t kInfoSize =
	    sizeof(chunksinfo_loopstart) + sizeof(chunksinfo_loopend) +
	    2 * sizeof(job_info::del_invalid) + 
            2 * sizeof(job_info::del_unused) + ...
...
```

lgsilva3087 replied:
Applied:

```cpp
uint32_t get_chunk_info_serialized_size() {
	// Some fields are multiplied by 2 because they are stored in both 'done' and 'not done' states.
	// See `chunk_store_info` function for details.
	constexpr uint32_t kInfoSize =
	    sizeof(chunksinfo_loopstart) + sizeof(chunksinfo_loopend) +
	    (2 * sizeof(job_info::del_invalid)) + (2 * sizeof(job_info::del_unused)) +
	    (2 * sizeof(job_info::del_diskclean)) + (2 * sizeof(job_info::del_overgoal)) +
	    (2 * sizeof(job_info::copy_undergoal)) + sizeof(loop_info::copy_rebalance);

	return kInfoSize;
}
```

dmga44 commented on src/master/restore.cc:
```diff
@@ -195,15 +195,16 @@
-	uint32_t inode;
+	inode_t inode;
 	EAT(ptr,filename,lv,'(');
 	GETU32(inode,ptr);
```
Please note this line `GETU32(inode,ptr);` (and its other occurrences in the file) needs to be changed in order to fully support 64 bits inodes.

lgsilva3087 replied:
Done. I created:
```cpp
#ifdef SAUNAFS_USE_INODE64
#define GETINODE(data,clptr) GETU64(data,clptr)
#else
#define GETINODE(data,clptr) GETU32(data,clptr)
#endif
```

dmga44 reviewed (commented):
Review part 2/TBD.

dmga44 commented on src/tools/set_trashtime.cc:
```diff
@@ -96,25 +111,25 @@ static int set_trashtime(const char *fname, uint32_t trashtime, uint8_t mode, in
-	changed = get32bit(&rptr);
-	notchanged = get32bit(&rptr);
-	notpermitted = get32bit(&rptr);
+	getINode(&rptr, changed);
+	getINode(&rptr, notchanged);
+	getINode(&rptr, notpermitted);
 	if ((mode & SMODE_RMASK) == 0) {
 		if (changed || mode == SMODE_SET) {
 			printf("%s: %" PRIu32 "\n", fname, trashtime);
```
Few lines down there are several calls to `print_number` function, the parameter `mode32` is set to 1, which means the numbers would be printed as `uint32_t` in some cases.

dmga44 replied:
Please note this comment could be applied in set_goal.cc file also.

lgsilva3087 replied:
Nice catch. Using now:
```cpp
// Used for printing numbers in tools
constexpr std::uint8_t kMode32 = kinode_t_size == 4 ? 1 : 0;
```

ralcolea commented on src/master/matoclserv.cc:
```diff
@@ -1636,13 +1783,23 @@ void matoclserv_fuse_register(matoclserventry *eptr,const uint8_t *data,uint32_t
+		constexpr uint32_t kBlobSizeWithRCode = kBlobSize + sizeof(rcode);
+		constexpr uint32_t kRegisterNewSessionMinSize =
+		    kBlobSizeWithRCode + sizeof(eptr->version) + sizeof(ileng) + sizeof(pleng);
+		constexpr uint32_t kRegisterNewMetaSessionMinSize =
```
Maybe `kRegisterNewSessionMinSize` and `kRegisterNewMetaSessionMinSize` can be switched and `kRegisterNewSessionMinSize` can reuse the value of `kRegisterNewSessionMinSize`: 
```cpp
kRegisterNewSessionMinSize = kRegisterNewMetaSessionMinSize + sizeof(pleng);
```

lgsilva3087 replied:
Applied

ralcolea reviewed (approved):
Great work!!! @lgsilva3087 :+1: :fire: :muscle:

lgsilva3087 committed:
fix(tools): Fix print_number calls in tools

The argument mode32 is used correctly now, according to the size of
inode_t.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
fix(master): Add GETINODE to restore changelogs

The macro GETU32 was used for inode_t variables. The new GETINODE was
added to conditionally use 32 or 64 bits according to the inode_t
definition.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
fix(all): Apply code review suggestions

Signed-off-by: guillex <guillex@leil.io>

dmga44 commented on src/mount/notification_area_logging.h:
```diff
@@ -148,7 +148,7 @@ inline void addNotificationMessage(const std::string &message) {
 
 inline void addPathByInodeBasedNotificationMessage(const std::string &message,
-                                                   uint32_t inode) {
+                                                   inode_t inode) {
 	if (stopThread || !gShowMessagesOnNotificationArea) { return; }
 	auto now = std::chrono::steady_clock::now();
 
```
The type of `fullPathFromInodeCache` needs to be changed as well.

lgsilva3087 replied:
Done

dmga44 commented on src/mount/readdata.cc:
```diff
@@ -571,7 +571,7 @@ void* read_worker(void *arg) {
 }
 
-ReadRecord *read_data_new(uint32_t inode) {
+ReadRecord *read_data_new(inode_t inode) {
```
Please note that the type of `gActiveReadRecords` needs to be changed also. From line 363:
- it is `using ReadRecords = std::unordered_multimap<uint32_t, ReadRecord *>;`
- it should be `using ReadRecords = std::unordered_multimap<inode_t, ReadRecord *>;`

lgsilva3087 replied:
Done.

dmga44 commented on src/mount/sauna_client.cc:
```diff
@@ -843,7 +843,7 @@ struct statvfs statfs(Context &ctx, Inode ino) {
 
-void access(Context &ctx, Inode ino, int mask) {
+void access(Context &ctx, inode_t ino, int mask) {
 	int status;
 
 	int mmode;
```
A couple of lines below the ones this comment is targetting, there is a call to `oplog_printf` that is printing some inode as it was (4 byte integer). Please consider updating those prints (there are lots of them).

lgsilva3087 replied:
Done (in many files).

dmga44 commented on src/mount/sauna_client.cc:
```diff
@@ -879,10 +879,10 @@ void access(Context &ctx, Inode ino, int mask) {
-EntryParam lookup(Context &ctx, Inode parent, const char *name) {
+EntryParam lookup(Context &ctx, inode_t parent, const char *name) {
 	EntryParam e;
 	uint64_t maxfleng;
-	uint32_t inode;
+	inode_t inode;
 	uint32_t nleng;
 	Attributes attr;
 	char attrstr[256];
```
The previously mentioned issue happens here a few lines below.

lgsilva3087 replied:
Done.

dmga44 commented on src/mount/special_lookup.cc:
```diff
@@ -47,7 +47,7 @@ static EntryParam lookup(const Context &ctx, Inode parent, const char *name,
 
 namespace InodeStats {
-static EntryParam lookup(const Context &ctx, Inode parent, const char *name,
+static EntryParam lookup(const Context &ctx, inode_t parent, const char *name,
 	                      char attrstr[256]) {
 	EntryParam e;
 	e.ino = inode_;
```
Please there are other `oplog_printf` issues like the ones pointed out before in this file (and other files).

lgsilva3087 replied:
Done

dmga44 commented on src/mount/writedata.cc:
```diff
@@ -2184,7 +2184,7 @@ void write_data_term(void) {
 	}
 }
 
-void *write_data_new(uint32_t inode) {
+void *write_data_new(inode_t inode) {
 	if (gUseInodeBasedWriteAlgorithm) { return InodeBasedWriteAlgorithm::write_data_new(inode); }
 	return ChunkBasedWriteAlgorithm::write_data_new(inode);
 }
```
Please apply analogous changes to the ones applid in the `InodeBasedWriteAlgorithm` to the `ChunkBasedWriteAlgorithm`. They all look good.

lgsilva3087 replied:
Done.

dmga44 commented on src/tools/quota_rep.cc:
```diff
@@ -67,7 +68,7 @@ static void quota_putc_plus_or_minus(uint64_t usage, uint64_t soft_limit, uint64
  */
 
-static void quota_print_entry(const std::string &path, uint32_t path_inode, int owner_type,
+static void quota_print_entry(const std::string &path, inode_t path_inode, int owner_type,
 							  uint32_t owner_id, const std::string &info,
 							  const QuotaDatabase::Limits &limit) {
 	static const char *owner_type_name[4] = {"User ", "Group", "Directory", "Unknown"};
```
On this topic there is the case when the quota is assigned to an inode (say a directory):
- in that case the the `owner_id` from this function could be referring to an inode.
- this leads us to the definition of `QuotaOwner` from src/protocol/quota.h
- the consequences of that change branch from that point on (to the master code and so on).

lgsilva3087 replied:
Done.

dmga44 reviewed (commented):
Review part 3/4.

lgsilva3087 committed:
fix(quotas): Use inode_t in QuotaOwner::ownerId

inode_t is the maximum expected type (types could be uid, gid and inodes).
Previously it was fixed uint32_t.

Signed-off-by: guillex <guillex@leil.io>

dmga44 commented on src/master/matoclserv.cc:
```diff
@@ -3271,7 +3733,7 @@ void matoclserv_fuse_settrashtime(matoclserventry *eptr, PacketHeader header, co
 }
 
 void matoclserv_fuse_getgoal(matoclserventry *eptr, PacketHeader header, const uint8_t *data) {
-	uint32_t inode;
+	inode_t inode;
 	uint32_t msgid;
 	uint8_t gmode;
 
```
In this function, the variables `fgtab` and `dgtab` are maps (goal -> number of files/directories with that goal). Though the way to calculate it is on the asked inode or recursively (which could be very expensive), and in the real use case it might not happen, we may have overflow here. Please note that the type of `legacyReplyFiles` and `legacyReplyDirectories` needs to changed accordingly if the suggestion is applied. The type of `FuseGetGoalStats` was already updated.

lgsilva3087 replied:
Done.

dmga44 reviewed (approved):
Review part 4/4. Please check my comment.

Amazing job @lgsilva3087 👍 🔥 🚀

lgsilva3087 committed:
fix(master): Use inode_t in GoalStatistics

The previous version was tied to uint32_t.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
fix(mount): Use inode_t in StatfsCacheKey

The commit was introduced later, so the rebase brought uint32_t instead
of inode_t.

Signed-off-by: guillex <guillex@leil.io>
